src/execution_plan.c

```diff
@@ -1258,60 +1417,129 @@ static Record* ExecutionPlan_AccumulateByKeyNextRecord(ExecutionPlan* ep, Execut
 static Record* ExecutionPlan_NextRecord(ExecutionPlan* ep, ExecutionStep* step, RedisModuleCtx* rctx){
     Record* r = NULL;
 
-    INIT_TIMER;
-    switch(step->type){
-    case READER:
-    	if(array_len(ep->errors) == 0){
-            GETTIME(&_ts);
-            ExecutionCtx ectx = ExecutionCtx_Initialize(rctx, ep);
-    	    r = step->reader.r->next(&ectx, step->reader.r->ctx);
-            GETTIME(&_te);
-    	    step->executionDuration += DURATION;
-    	    if(!r && ectx.err){
-    	        r = RG_ErrorRecordCreate(ectx.err, strlen(ectx.err) + 1);
-    	    }
-    	}
-        break;
-    case MAP:
-    	r = ExecutionPlan_MapNextRecord(ep, step, rctx);
-    	break;
-    case FLAT_MAP:
-    	r = ExecutionPlan_FlatMapNextRecord(ep, step, rctx);
-    	break;
-    case FILTER:
-    	r = ExecutionPlan_FilterNextRecord(ep, step, rctx);
-    	break;
-    case EXTRACTKEY:
-    	r = ExecutionPlan_ExtractKeyNextRecord(ep, step, rctx);
-    	break;
-    case GROUP:
-    	r = ExecutionPlan_GroupNextRecord(ep, step, rctx);
-    	break;
-    case REDUCE:
-    	r = ExecutionPlan_ReduceNextRecord(ep, step, rctx);
-    	break;
-    case REPARTITION:
-    	r = ExecutionPlan_RepartitionNextRecord(ep, step, rctx);
-    	break;
-    case COLLECT:
-    	r = ExecutionPlan_CollectNextRecord(ep, step, rctx);
-    	break;
-    case FOREACH:
-    	r = ExecutionPlan_ForEachNextRecord(ep, step, rctx);
-    	break;
-    case LIMIT:
-    	r = ExecutionPlan_LimitNextRecord(ep, step, rctx);
-    	break;
-    case ACCUMULATE:
-    	r = ExecutionPlan_AccumulateNextRecord(ep, step, rctx);
-    	break;
-    case ACCUMULATE_BY_KEY:
-    	r = ExecutionPlan_AccumulateByKeyNextRecord(ep, step, rctx);
-		break;
-    default:
-        RedisModule_Assert(false);
+    if(step->isDone){
         return NULL;
     }
+
+    INIT_TIMER;
+
+    bool isEmpty;
+
+    if(step->pendingCtx){
+        r = ExecutionPlan_GetRecordFromPending(step->pendingCtx, &isEmpty);
+        if(r){
+            return r;
+        }
+        RedisModule_Assert(isEmpty);
+        ExecutionPlan_PendingCtxFreeInternals(step->pendingCtx);
+        step->pendingCtx = NULL;
+    }
+
+    StepPendingCtx** pendingCtxs = ep->pendingCtxs + step->stepId;
+
+    if(*pendingCtxs){
+        r = ExecutionPlan_GetRecordFromPending(*pendingCtxs, &isEmpty);
+        if(r){
+            return r;
+        }
+    }
+
+    r = &DummyRecord;
+    while(r == &DummyRecord &&
+            (!(*pendingCtxs) || Gears_listLength((*pendingCtxs)->records) < (*pendingCtxs)->maxSize)){
+
+        switch(step->type){
```

Comment:
consider maybe having a function pointer in the step struct, to point to the step execution function.

Reply:
You potentially lose performance, function pointers can not be inlined and slower to call because of reference locality.